h2/src/main/org/h2/engine/Session.java

```diff
@@ -681,13 +681,16 @@ private void markUsedTablesAsUpdated() {
     }
 
     private void analyzeTables() {
+        // take a local copy and clear because in rare cases we can call
+        // back into markTableForAnalyze while iterating here
+        HashSet<Table> tablesToAnalyzeLocal = tablesToAnalyze;
+        tablesToAnalyze = null;
         int rowCount = getDatabase().getSettings().analyzeSample / 10;
-        for (Table table : tablesToAnalyze) {
+        for (Table table : tablesToAnalyzeLocal) {
```

Comment:
If it is unprotected by lock, and we assume it is, because of CME, it will create race condition still, since tablesToAnalyze is a non-volatile field.

Reply:
ConcurrentModificationException in this specific case does not imply a threading issue - what it means is that a collection was modified while an iterator was iterating over the collection

Reply:
Of course, but since stack trace does not show any re-entrance, one would assume modification from another thread.